clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -528,20 +527,37 @@ object GarbageCollector {
   ) = {
     val reportLogsDst = concatToGCLogsPrefix(storageNSForHadoopFS, "summary")
     val reportExpiredDst = concatToGCLogsPrefix(storageNSForHadoopFS, "expired_addresses")
+    val deletedObjectsDst = concatToGCLogsPrefix(storageNSForHadoopFS, "deleted_objects")
+
     val time = DateTimeFormatter.ISO_INSTANT.format(java.time.Clock.systemUTC.instant())
     writeParquetReport(commitsDF, reportLogsDst, time, "commits.parquet")
-    writeParquetReport(expiredAddresses, reportExpiredDst, time)
-    writeJsonSummary(configMapper, reportLogsDst, removed.count(), gcRules, time)
-
-    removed
-      .withColumn(MARK_ID_KEY, lit(markID))
-      .withColumn(RUN_ID_KEY, lit(runID))
-      .write
-      .partitionBy(MARK_ID_KEY, RUN_ID_KEY)
-      .mode(SaveMode.Overwrite)
-      .parquet(
-        concatToGCLogsPrefix(storageNSForHadoopFS, s"deleted_objects/$time/deleted.parquet")
-      )
+    try {
+      writeParquetReport(expiredAddresses, reportExpiredDst, time)
+      val expiredDF = spark.read.parquet(f"${reportExpiredDst}/dt=${time}/")
+      println(f"Total expired addresses: ${expiredDF.count()}")
+    } catch {
+      case e: Throwable => {
+        println("Error when trying to get expired addresses count, moving on:")
+        e.printStackTrace()
+      }
+    }
+    try {
+      val removedCount = removed.count()
+      writeJsonSummary(configMapper, reportLogsDst, removedCount, gcRules, time)
+      removed
+        .withColumn(MARK_ID_KEY, lit(markID))
+        .withColumn(RUN_ID_KEY, lit(runID))
+        .write
+        .partitionBy(MARK_ID_KEY, RUN_ID_KEY)
+        .mode(SaveMode.Overwrite)
+        .parquet(f"${deletedObjectsDst}/dt=$time")
+      println(f"Total objects deleted (or already had been deleted): ${removedCount}")
```

Comment:
>or already had been deleted

Isn't the removed count represent the exact number of objects that were deleted? I mean, given that gc is not incremental I expect expired addresses to include addresses of already deleted objects but for removed objects I expect is to be reliable for the run.

Reply:
The count represent all the expired addresses that, at the end of this run, were not present in the storage. It does not know if they were deleted in this run, in previous runs, or by anyone else for that matter.

IMO, it's good that we are being consistent with how the storage reports deletion. The only problem here is that it may not be consistent with Azure (I'm not sure whether it is or not).